src/ffi.rs

```diff
@@ -578,6 +578,199 @@ pub unsafe extern "C" fn context_get_result(
+/// If `operators` is non-null, it will write the used operators with bitflags to the provided pointer.
+/// The bitflags is defined by `BinaryOperatorFlags` and it must not contain any bits from `BinaryOperatorFlags::UNUSED`.
+///
+///
+/// # Panics
+///
+/// This function will panic when:
+///
+/// - `atc` doesn't point to a valid C-style string.
+/// - `fields_len` and `fields_total` are null when `fields_buf` is non-null.
+///
+/// # Safety
+///
+/// Violating any of the following constraints will result in undefined behavior:
+///
+/// - `atc` must be a valid pointer to a C-style string, must be properly aligned,
+///    and must not have '\0' in the middle.
+/// - `schema` must be a valid pointer returned by [`schema_new`].
+/// - `fields_buf` must be a valid to write for `fields_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_len` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_total` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `operators` must be a valid to write for `size_of::<u64>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `errbuf` must be valid to read and write for `errbuf_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned.
+/// - `errbuf_len` must be valid to read and write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned.
+/// - If `fields_buf` is non-null, `fields_len` and `fields_total` must be non-null.
+/// - If `fields_buf` is null, `fields_len` and `fields_total` can be non-null
+///   for writing required buffer length and total number of fields.
+#[cfg(feature = "expr_validation")]
+#[no_mangle]
+pub unsafe extern "C" fn expression_validate(
+    atc: *const u8,
+    schema: &Schema,
+    fields_buf: *mut u8,
+    fields_len: *mut usize,
+    fields_total: *mut usize,
+    operators: *mut u64,
+    errbuf: *mut u8,
+    errbuf_len: *mut usize,
+) -> i64 {
+    use std::collections::HashMap;
+
+    use crate::ast::{BinaryOperatorFlags, Expression, LogicalExpression};
+    use crate::parser::parse;
+    use crate::semantics::FieldCounter;
+    use crate::semantics::Validate;
+
+    let atc = ffi::CStr::from_ptr(atc as *const c_char).to_str().unwrap();
+    let errbuf = from_raw_parts_mut(errbuf, ERR_BUF_MAX_LEN);
+
+    // Parse the expression
+    let result = parse(atc).map_err(|e| e.to_string());
+    if let Err(e) = result {
+        let errlen = min(e.len(), *errbuf_len);
+        errbuf[..errlen].copy_from_slice(&e.as_bytes()[..errlen]);
+        *errbuf_len = errlen;
+        return ATC_ROUTER_EXPRESSION_VALIDATE_FAILED;
+    }
+    // Unwrap is safe since we've already checked for error
+    let ast = result.unwrap();
+
+    // Validate expression with schema
+    if let Err(e) = ast.validate(schema).map_err(|e| e.to_string()) {
```

Comment:
Maybe we could wrap `parse`, `validate` and `add_to_counter` logic in a single function, so that it could be call for both `Router` and `expression_validate`. 
The reason I mention this is that there maybe data structure change for above functions(e.g. I may change `add_to_counter`'s parameter in my PR for indexed fields), wrapping the implementaion logic in a single function would bring the benefit that we don't need to change every caller's code in future.

Reply:
I just noticed that here we use `ast.add_to_counter`, and when Hai's PR merged, router will use `cir.add_to_counter`, there would be different. Maybe not easy to extract to general logic.

Reply:
Given that `validate` is provided by a trait and `AST` should always implement this (I don't think we want to perform the validation on other data structures.) and `parse` is very natural from `AST` from `String`, I think `parse` and `validate` should be considered very stable interfaces. Indeed calling `parse` and then `validate` is a widespread use case but I don't think we should add a new interface just calling them together.

However, I did realize that we should not depend on `add_to_counter` for this interface thanks for your comment. Because `counter` should be an internal underlying method to achieve something that should not be public as the interface. Hai's PR changed the `semantics::FieldCounter` to be implemented by `Cir` rather than `AST` because of that "internality". This is an interface change and not a data structure change.

I'd like to introduce two traits like `semantics::GetFields` and `semantics::GetOperators` and implement this for `AST` only. This may provide a more idiomatic way to get data from `AST`.

WDYT? @dndx @ADD-SP @ProBrian

Reply:
> I'd like to introduce two traits like `semantics::GetFields` and `semantics::GetOperators` and implement this for `AST` only. This may provide a more idiomatic way to get data from `AST`.

Sounds good, that will make the FFI layer more clear and brief.